sbi/diagnostics/lc2st.py

```diff
@@ -0,0 +1,786 @@
+                    "early_stopping": True,
+                    "n_iter_no_change": 50,
+                }
+        elif "random_forest" in classifier.lower():
+            self.clf_class = RandomForestClassifier
+            if clf_kwargs is None:
+                self.clf_kwargs = {}
+        elif "custom":
+            if clf_class is None or clf_kwargs is None:
+                raise ValueError(
+                    "Please provide a valid sklearn classifier class and kwargs."
+                )
+            self.clf_class = clf_class
+            self.clf_kwargs = clf_kwargs
+        else:
+            raise NotImplementedError
+
+        # initialize classifiers, will be set after training
+        self.trained_clfs = None
+        self.trained_clfs_null = None
+
+        # parameters for the null hypothesis testing
+        self.num_trials_null = num_trials_null
+        self.permutation = permutation
+        # can be specified if known and independent of x (see `LC2ST-NF`)
+        self.null_distribution = None
+
+    def _train(
+        self,
+        theta_p: Tensor,
+        theta_q: Tensor,
+        x_p: Tensor,
+        x_q: Tensor,
+        verbosity: int = 0,
+    ) -> List[Any]:
+        """Returns the classifiers trained on observed data.
+
+        Args:
+            theta_p: Samples from P, of shape (sample_size, dim).
+            theta_q: Samples from Q, of shape (sample_size, dim).
+            x_p: Observations corresponding to P, of shape (sample_size, dim_x).
+            x_q: Observations corresponding to Q, of shape (sample_size, dim_x).
+            verbosity: Verbosity level, defaults to 0.
+
+        Returns:
+            List of trained classifiers for each cv fold.
+        """
+
+        # prepare data
+
+        if self.z_score:
+            theta_p = (theta_p - self.theta_p_mean) / self.theta_p_std
+            theta_q = (theta_q - self.theta_p_mean) / self.theta_p_std
+            x_p = (x_p - self.x_p_mean) / self.x_p_std
+            x_q = (x_q - self.x_p_mean) / self.x_p_std
+
+        # initialize classifier
+        clf = self.clf_class(**self.clf_kwargs or {})
+
+        # cross-validation
+        if self.num_folds > 1:
+            trained_clfs = []
+            kf = KFold(n_splits=self.num_folds, shuffle=True, random_state=self.seed)
+            cv_splits = kf.split(theta_p)
+            for train_idx, _ in tqdm(
+                cv_splits, desc="Cross-validation", disable=verbosity < 1
+            ):
+                # get train split
+                theta_p_train, theta_q_train = theta_p[train_idx], theta_q[train_idx]
+                x_p_train, x_q_train = x_p[train_idx], x_q[train_idx]
+
+                # train classifier
+                clf_n = train_lc2st(
+                    theta_p_train, theta_q_train, x_p_train, x_q_train, clf
+                )
+
+                trained_clfs.append(clf_n)
+
+        # ensembling
+        if self.num_ensemble > 1:
+            trained_clfs = []
```

Comment:
if both `self.num_folds` and `self.num_ensemble` are  `>1` then this effectively resets `trained_clfs = []` and discards the trained classifiers from above, no?

I am wondering what is the difference between doing a k-fold CV and ensembling (except that CV takes different train-test splits?)  do we really need the ensembling here?

Reply:
Oh you might be right! Let me check! 

Ensembling reduces the variance coming from the classifier initialization (with different seeds). I stubled upon an example where this is quite important.

Reply:
I'm just wondering if we should allow to do CV and ensembling together (i.e. ensembling inside the CV loop) or if this should raise an error.

Reply:
maybe we can just add seeding to the CV? This then should be the same as doing ensembling except that CV iterates on different test and train splits?

Reply:
Yes, but for small datasets this might not be appropriate. I don't know... Maybe @psteinb can give his opinion on that?

Reply:
cross val makes different data splits so this would need to be integrated in the null hypothesis I think...